packages/ejson/lib/src/decoding.dart

```diff
@@ -0,0 +1,272 @@
+    Map<dynamic, dynamic> m => m.map((key, value) => MapEntry(key as K, fromEJson<V>(value))),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+double _decodeDouble(EJsonValue ejson) {
+  return switch (ejson) {
+    double d => d, // relaxed mode
+    //{'\$numberDouble': double d} => d,
+    {'\$numberDouble': String s} => switch (s) {
+        'NaN' => double.nan,
+        'Infinity' => double.infinity,
+        '-Infinity' => double.negativeInfinity,
+        _ => double.tryParse(s) ?? raiseInvalidEJson(ejson),
+      },
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+int _decodeInt(EJsonValue ejson) {
+  return switch (ejson) {
+    int i => i, // relaxed mode
+    {'\$numberInt': String i} => int.tryParse(i) ?? raiseInvalidEJson(ejson),
+    {'\$numberLong': String i} => int.tryParse(i) ?? raiseInvalidEJson(ejson),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+Key _decodeKey(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$minKey': 1} => Key.min,
+    {'\$maxKey': 1} => Key.max,
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+// ignore: prefer_void_to_null
+Null _decodeNull(EJsonValue ejson) {
+  return switch (ejson) {
+    null => null,
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+T? _decodeNullable<T>(EJsonValue ejson) {
+  if (ejson == null) {
+    return null;
+  }
+  return fromEJson<T>(ejson);
+}
+
+num _decodeNum(EJsonValue ejson) {
+  return switch (ejson) {
+    num n => n, // relaxed mode
+    {'\$numberLong': _} => _decodeInt(ejson),
+    {'\$numberInt': _} => _decodeInt(ejson),
+    {'\$numberDouble': _} => _decodeDouble(ejson),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+ObjectId _decodeObjectId(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$oid': String s} => ObjectId.fromHexString(s),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+String _decodeString(EJsonValue ejson) {
+  return switch (ejson) {
+    String s => s,
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+Symbol _decodeSymbol(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$symbol': String s} => Symbol(s),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+Undefined<T> _decodeUndefined<T>(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$undefined': 1} => Undefined<T>(),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+UndefinedOr<T> _decodeUndefinedOr<T>(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$undefined': 1} => Undefined<T>(),
+    _ => _decodeDefined(ejson),
+  };
+}
+
+Uuid _decodeUuid(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$binary': {'base64': String s, 'subType': '04'}} => Uuid.fromBytes(base64.decode(s).buffer),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+Uint8List _decodeBinary(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$binary': {'base64': String s, 'subType': _}} => base64.decode(s),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+/// Thrown when a value cannot be decoded from [ejson].
+class InvalidEJson implements Exception {
+  final EJsonValue ejson;
+  final Type type;
```

Comment:
Should we have docs here?

Reply:
Should we? I guess we should.. will fix.